File: client_2finance/wallet.go
package client_2finance

import (

	"fmt"
	"gitlab.com/2finance/2finance-network/blockchain/encryption/keys"
	"gitlab.com/2finance/2finance-network/blockchain/types"
	"gitlab.com/2finance/2finance-network/blockchain/contract/walletV1"

	"gitlab.com/2finance/2finance-network/blockchain/utils"

)
// AddWallet creates a new wallet
// and sends a transaction to the network
// Amount is a string representation of the amount to be added
// If amount is empty, it defaults to "0"
// Amount must be sent considering decimals
// For example, if the amount is 100000, and decimals is 18, the amount in database will be 100000000000000000000000
// if the amonut is 0,0000000001, and decimals is 18, the amount in database will be 100000000
func (c *networkClient) AddWallet(address, pubKey string) (types.ContractOutput, error) {
	if pubKey == "" {
		return types.ContractOutput{}, fmt.Errorf("public key not set")
	}
	if err := keys.ValidateEDDSAPublicKey(pubKey); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid public key: %w", err)
	}
	if address == "" {
		return types.ContractOutput{}, fmt.Errorf("contract address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(address); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid contract address: %w", err)
	}

	from := c.publicKey
	if from == "" {
		return types.ContractOutput{}, fmt.Errorf("from address not set")
	}

	to := address
	method := walletV1.METHOD_ADD_WALLET
	data := map[string]interface{}{
		"address":    address,
		"public_key": pubKey,
		//TODO REMOVER
		"amount":     "0",
	}

	contractOutput, err := c.SignAndSendTransaction(
		from,
		to,
		method,
		data)
	if err != nil {
		return types.ContractOutput{}, fmt.Errorf("failed to send transaction: %w", err)
	}

	return contractOutput, nil
}

func (c *networkClient) GetWallet(pubKey string) (types.ContractOutput, error) {
	
	if pubKey == "" {
		return types.ContractOutput{}, fmt.Errorf("public key not set")
	}
	err := keys.ValidateEDDSAPublicKey(pubKey)
	if err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid public key: %w", err)
	}

	method := walletV1.METHOD_GET_WALLET_BY_PUBLIC_KEY
	data := map[string]interface{}{
		"public_key": pubKey,
	}

	contractOutput, err := c.GetState("", method, data)
	if err != nil {
		return types.ContractOutput{}, fmt.Errorf("failed to get state: %w", err)
	}

	return contractOutput, nil
}

func (c *networkClient) TransferWallet(to, amount string, decimals int) (types.ContractOutput, error) {
	if c.publicKey == "" {
		return types.ContractOutput{}, fmt.Errorf("public key not set")
	}
	if to == "" {
		return types.ContractOutput{}, fmt.Errorf("to address not set")
	}
	if to == c.publicKey {
		return types.ContractOutput{}, fmt.Errorf("cannot transfer to the same address")
	}
	if err := keys.ValidateEDDSAPublicKey(to); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid to address: %w", err)
	}
	if amount == "" {
		return types.ContractOutput{}, fmt.Errorf("amount not set")
	}

	if decimals != 0 {
		amountConverted, err := utils.RescaleDecimalString(amount, 0, decimals)
		if err != nil {
			return types.ContractOutput{}, fmt.Errorf("failed to convert amount to big int: %w", err)
		}
		amount = amountConverted
	}

	method := walletV1.METHOD_TRANSFER_WALLET
	data := map[string]interface{}{
		"from":    c.publicKey,
		"to":      to,
		"amount":  amount,
	}

	contractOutput, err := c.SignAndSendTransaction(
		c.publicKey,
		to,
		method,
		data)
	if err != nil {
		return types.ContractOutput{}, fmt.Errorf("failed to send transaction: %w", err)
	}
	
	//TODOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
	// transferOutput.Transfer = contractOutput.States[0].Object.(*domain.Transfer)
	// transferOutput.EventTransfer = &contractOutput.Events[0]
	// transferOutput.LogTypeTransfer = &contractOutput.LogTypes[0]
	
	// transferOutput.WalletSender = contractOutput.States[1].Object.(*domain.Wallet)
	// transferOutput.EventSender = &contractOutput.Events[1]
	// transferOutput.LogTypeSender = &contractOutput.LogTypes[1]

	// transferOutput.WalletReceiver = contractOutput.States[2].Object.(*domain.Wallet)
	// transferOutput.EventReceiver = &contractOutput.Events[2]
	// transferOutput.LogTypeReceiver = &contractOutput.LogTypes[2]

	return contractOutput, nil
}

File: dart/blockchain/lib/wallet.dart
part of two_finance_blockchain;

extension Wallet on TwoFinanceBlockchain {
  Future<ContractOutput> addWallet(String address, String pubKey) async {
    if (address.isEmpty) {
      throw ArgumentError('contract address not set');
    }
    if (pubKey.isEmpty) {
      throw ArgumentError('public key not set');
    }

    final from = publicKeyHex ?? '';
    if (from.isEmpty) {
      throw ArgumentError('from address not set');
    }

    KeyManager.validateEDDSAPublicKeyHex(from);
    KeyManager.validateEDDSAPublicKeyHex(address);
    KeyManager.validateEDDSAPublicKeyHex(pubKey);

    final chainID = _chainID;
    final String to = address;
    const String method = METHOD_ADD_WALLET;
    const int version = 1;
    final String uuid7 = newUUID7();

    final JsonMessage data = {'address': address, 'public_key': pubKey};

    try {
      final contractOutput = await signAndSendTransaction(
        chainID: chainID,
        from: from,
        to: to,
        method: method,
        data: data,
        version: version,
        uuid7: uuid7,
      );
      return contractOutput;
    } catch (e) {
      throw Exception('failed to send transaction: $e');
    }
  }

  Future<ContractOutput> getWalletByPublicKey(String pubKey) async {
    if (pubKey.isEmpty) {
      throw ArgumentError('public key not set');
    }

    try {
      KeyManager.validateEDDSAPublicKeyHex(pubKey);
    } catch (e) {
      throw ArgumentError('invalid public key: $e');
    }

    const String method = METHOD_GET_WALLET_BY_PUBLIC_KEY;
    const String contractVersion = WALLET_CONTRACT_V2;
    final JsonMessage data = {
      'public_key': pubKey,
      'contract_version': contractVersion,
    };

    try {
      final contractOutput = await getState(to: '', method: method, data: data);
      return contractOutput;
    } catch (e) {
      throw Exception('failed to get state: $e');
    }
  }

  Future<ContractOutput> getWalletByAddress(String address) async {
    if (address.isEmpty) {
      throw ArgumentError('contract address not set');
    }

    KeyManager.validateEDDSAPublicKeyHex(address);

    const String method = METHOD_GET_WALLET_BY_ADDRESS;
    final JsonMessage data = {'address': address};

    try {
      final contractOutput = await getState(
        to: address,
        method: method,
        data: data,
      );
      return contractOutput;
    } catch (e) {
      throw Exception('failed to get state: $e');
    }
  }
}
